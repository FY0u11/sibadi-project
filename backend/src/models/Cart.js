const { sequelize, Sequelize: DataTypes } = require('../config/db')

class Cart extends DataTypes.Model {}

Cart.init({
  userId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: {
      min: 1
    }
  },
  productId: {
    type: DataTypes.INTEGER,
    allowNull: false,
    validate: {
      min: 1
    }
  }
}, {
  sequelize,
  modelName: 'cart'
})

module.exports = Cart
