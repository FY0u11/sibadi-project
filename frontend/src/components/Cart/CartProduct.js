import React from 'react'

import Button from '../Button'
import Card from '../Card'

const CartProduct = ({ product, onRemove }) => {
  const image = <img src={`/uploads/${product.imagePath}`} className="card-img" alt="..." />
  const body = (
    <div>
      <Button action={onRemove}>Remove</Button>
      <span>{product.price} $</span>
    </div>
  )

  return <Card image={image} title={product.name} body={body} width="400px" />
}

export default CartProduct
