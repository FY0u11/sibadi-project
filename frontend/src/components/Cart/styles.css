.my-cart {
  position: fixed;
  top: 56px;
  right: 20px;
  padding: 20px;
  max-height: 400px;
  border-radius: 0 0 10px 10px;
  overflow: scroll;
  overflow-x: hidden;
  z-index: 2;
}
