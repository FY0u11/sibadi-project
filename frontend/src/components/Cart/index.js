import React from 'react'
import { connect } from 'react-redux'

import { createOrderRequest, clearCartRequest } from '../../utils/api'
import { setUser } from '../../store/actions'
import { setCartProducts } from '../../store/components/cart/actions'
import { setOrders } from '../../store/scenes/orders/actions'
import {
  setNotificationError,
  setNotificationSuccess
} from '../../store/components/notifications/actions'

import Cart from './Cart'

const CartContainer = ({
  user,
  setUser,
  orders,
  setOrders,
  isCartShown,
  cartProducts,
  setCartProducts,
  setNotificationError,
  setNotificationSuccess
}) => {
  const clearCart = async () => {
    try {
      await clearCartRequest()
      setCartProducts([])
    } catch (error) {
      setNotificationError(error.message)
    }
  }

  const buy = async () => {
    try {
      const { order, newBudget } = await createOrderRequest()
      setUser({ ...user, budget: newBudget })
      await clearCart()
      setOrders([...orders, order])
      setNotificationSuccess(`Your order #${order.id} successfully accepted`)
    } catch (error) {
      setNotificationError(error.message)
    }
  }

  const onRemove = async id => {
    try {
      await clearCartRequest({ productId: id })
      setCartProducts(cartProducts.filter(cartProduct => cartProduct.id !== id))
    } catch (error) {
      setNotificationError(error.message)
    }
  }

  return isCartShown && cartProducts.length ? (
    <Cart cartProducts={cartProducts} clearCart={clearCart} onRemove={onRemove} buy={buy} />
  ) : null
}

const putStateToProps = state => ({
  user: state.root.user,
  orders: state.orders.orders,
  isCartShown: state.cart.isCartShown,
  cartProducts: state.cart.cartProducts
})

const putActionsToProps = {
  setUser,
  setOrders,
  setCartProducts,
  setNotificationError,
  setNotificationSuccess
}

export default connect(putStateToProps, putActionsToProps)(CartContainer)
