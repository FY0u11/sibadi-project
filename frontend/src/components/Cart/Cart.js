import React from 'react'

import './styles.css'
import CartProduct from './CartProduct'
import Button from '../Button'

const Cart = ({ cartProducts, onRemove, clearCart, buy }) => {
  const totalPrice = cartProducts.reduce((sum, product) => sum + product.price, 0)

  return (
    <div className="my-cart bg-light">
      <h3>Total: {totalPrice}</h3>
      {cartProducts.map(product => (
        <CartProduct key={product.id} product={product} onRemove={() => onRemove(product.id)} />
      ))}
      <div style={{ marginBottom: '20px' }}>
        <Button action={buy}>Buy</Button>
        <Button action={clearCart}>Clear cart</Button>
      </div>
    </div>
  )
}

export default Cart
