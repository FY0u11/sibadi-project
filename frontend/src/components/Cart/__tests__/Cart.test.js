import React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import Cart from '../Cart'

afterEach(cleanup)

describe('Testing Cart component', () => {
  it('It should be properly rendered', () => {
    const cartProducts = [
      { id: 2, price: 600, name: 'prodcut name 1', imagePath: 'path/1' },
      { id: 42, price: 66, name: 'prodcut name 2', imagePath: 'path/1' }
    ]
    const { asFragment } = render(<Cart cartProducts={cartProducts} />)
    expect(asFragment()).toMatchSnapshot()
  })
  it('The callback function should be called with argument 42', () => {
    const cartProducts = [{ id: 42 }]
    const callback = jest.fn()
    const { getByText } = render(<Cart cartProducts={cartProducts} onRemove={callback} />)
    const removeButton = getByText('Remove')
    fireEvent.click(removeButton)
    expect(callback).toBeCalledTimes(1)
    expect(callback).toBeCalledWith(42)
  })
})
