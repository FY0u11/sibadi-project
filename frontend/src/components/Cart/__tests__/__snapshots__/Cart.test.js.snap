// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Cart component It should be properly rendered 1`] = `
<DocumentFragment>
  <div
    class="my-cart bg-light"
  >
    <h3>
      Total: 666
    </h3>
    <div
      class="card mb-3 my-card"
      style="margin: 0px auto; width: 400px;"
    >
      <div
        class="row no-gutters"
      >
        <div
          class="col-md-4"
        >
          <img
            alt="..."
            class="card-img"
            src="/uploads/path/1"
          />
        </div>
        <div
          class="col-md-8"
        >
          <div
            class="card-body"
          >
            <h5
              class="card-title"
            >
              prodcut name 1
            </h5>
            <div>
              <button
                class="btn btn-dark"
                style="margin-right: 20px;"
              >
                Remove
              </button>
              <span>
                600 $
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="card mb-3 my-card"
      style="margin: 0px auto; width: 400px;"
    >
      <div
        class="row no-gutters"
      >
        <div
          class="col-md-4"
        >
          <img
            alt="..."
            class="card-img"
            src="/uploads/path/1"
          />
        </div>
        <div
          class="col-md-8"
        >
          <div
            class="card-body"
          >
            <h5
              class="card-title"
            >
              prodcut name 2
            </h5>
            <div>
              <button
                class="btn btn-dark"
                style="margin-right: 20px;"
              >
                Remove
              </button>
              <span>
                66 $
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      style="margin-bottom: 20px;"
    >
      <button
        class="btn btn-dark"
        style="margin-right: 20px;"
      >
        Buy
      </button>
      <button
        class="btn btn-dark"
        style="margin-right: 20px;"
      >
        Clear cart
      </button>
    </div>
  </div>
</DocumentFragment>
`;
